autoendpoint/src/db/client.rs

```diff
@@ -0,0 +1,184 @@
+                retryable_delete_error(self.metrics.clone()),
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Get the set of channel IDs for a user
+    pub async fn get_user_channels(&self, uaid: Uuid) -> DbResult<HashSet<Uuid>> {
+        // Channel IDs are stored in a special row in the message table, where
+        // chidmessageid = " "
+        let input = GetItemInput {
+            table_name: self.message_table.clone(),
+            consistent_read: Some(true),
+            key: ddb_item! {
+                uaid: s => uaid.to_simple().to_string(),
+                chidmessageid: s => " ".to_string()
+            },
+            ..Default::default()
+        };
+
+        let output = retry_policy()
+            .retry_if(
+                || self.ddb.get_item(input.clone()),
+                retryable_getitem_error(self.metrics.clone()),
+            )
+            .await?;
+
+        // The channel IDs are in the notification's `chids` field
+        let channels = output
+            .item
+            // Deserialize the notification
+            .map(serde_dynamodb::from_hashmap::<DynamoDbNotification, _>)
+            .transpose()?
+            // Extract the channel IDs
+            .and_then(|n| n.chids)
+            .unwrap_or_default();
+
+        // Convert the IDs from String to Uuid
+        let channels = channels
+            .into_iter()
+            .map(|s| Uuid::parse_str(&s))
+            .filter_map(Result::ok)
+            .collect();
+
+        Ok(channels)
+    }
+
+    /// Remove the node ID from a user in the router table.
+    /// The node ID will only be cleared if `connected_at` matches up with the
+    /// item's `connected_at`.
+    pub async fn remove_node_id(
+        &self,
+        uaid: Uuid,
+        node_id: String,
+        connected_at: u64,
+    ) -> DbResult<()> {
+        let update_item = UpdateItemInput {
+            key: ddb_item! { uaid: s => uaid.to_simple().to_string() },
+            update_expression: Some("REMOVE node_id".to_string()),
+            condition_expression: Some("(node_id = :node) and (connected_at = :conn)".to_string()),
+            expression_attribute_values: Some(hashmap! {
+                ":node".to_string() => val!(S => node_id),
+                ":conn".to_string() => val!(N => connected_at.to_string())
+            }),
+            table_name: self.router_table.clone(),
+            ..Default::default()
+        };
+
+        retry_policy()
+            .retry_if(
+                || self.ddb.update_item(update_item.clone()),
+                retryable_updateitem_error(self.metrics.clone()),
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    /// Store a single message
+    pub async fn store_message(&self, uaid: Uuid, message: Notification) -> DbResult<()> {
+        let put_item = PutItemInput {
+            item: serde_dynamodb::to_hashmap(&DynamoDbNotification::from_notif(&uaid, message))
+                .unwrap(),
```

Comment:
Should we have this return a proper error rather than a panic? (In case someone gives us a TTL of `None` or something stupid?) Could probably have that return a 400 in those cases, which might help the abuse filters.

Reply:
> Serialization can fail if T's implementation of Serialize decides to fail, or if T contains a map with non-string keys.

https://docs.rs/serde_dynamodb/0.5.1/serde_dynamodb/fn.to_hashmap.html

I don't think either of these cases can occur here, but in future PRs I did change `DbError` to support serialization in addition to deserialization, so I will backport that change and handle the error here.